### PR TITLE
Multiple General Improvements

### LINKED ISSUE #81: Correct scaling for MVP when resizing window
Currentl the background is not drawn properly and UI elements are not rendere correctly when resizing the window.

### LINKED ISSUE #114: [MVP] Improved ship controls
Add cooldowns to attacks, movement, etc.
Add modifiers that affects these cooldowns, movement speed etc.

### PR DESCRIPTION
**Pull request by JoXW100**
Fixes #81 
Fixes #114 

- Implemented `NavMesh` to store navigation data between pathfinding.
- Implemented `TresureIsland` gameobject and spawned them dynamically 
- Made `Base` into a type of ship to simplify targeting
- Added spawn locations to `ShipInfo` to reduce duplication of hard-coded positions, and to support multiple spawns
- Fixed error in the win screen causing crash
- Improved pathfinding
- Added ship debug view of the navmesh.
- Made `DataObject` methods and constructors into `constexpr`
- Added text scaling

From #120 
- Improved server tickrate
- Added cooldowns to ship movement and attacks affected by their speed.

### CODE REVIEW

**Comment by joelsiks** [comment]
Just want to add that some of the changes here could have been made into their own pull requests since this one is very large (42 files changed). For example, the TreasureIsland extraction could be a separate issue, or at the very least separate commits.

**Comment by JoXW100** [comment]
> Just want to add that some of the changes here could have been made into their own pull requests since this one is very large (42 files changed). For example, the TreasureIsland extraction could be a separate issue, or at the very least separate commits.

I agree with you and have made this draft into a pr for general improvements as the actual combat elements are yet to be implemented. The combat implementation itself will follow in a separate pr.

**Review by joelsiks — CHANGES_REQUESTED** [review]
See comments. You proably forgot to update the doc comment for `Scene::BuildNavMesh`. Otherwise good stuff. I've tried running and it seems to work fine. I have a few small things that would be easiest to show later, but it relates to moving ship(s).

File: src/engine/IDisplayLayer.hpp [hunk 1]
```diff
@@ -23,11 +23,14 @@ class IDisplayLayer {
...
     virtual void RemoveDisplayable(const std::string &identifier);
     virtual bool ExistsDisplayable(const std::string &identifier);
+    virtual std::shared_ptr<IInteractiveDisplayable>
+    FindDisplayable(const std::string &identifier);
 
     inline size_t NumDisplayables() const { return m_displayables.Size(); }
     std::vector<std::string> GetDisplayableNames();
 
     void RebuildQuadTree(const Vector2 &windowSize);
+    inline const QuadTree &GetQuadTree() const { return m_quadtree; }
```
[the remark this answers is not in the document]
Comment (JoXW100) [reply src/engine/IDisplayLayer.hpp:33]: That is why a `const` reference is returned, it cannot be altered.

**Review by caspernorrbin — APPROVED** [review]

File: src/engine/Scene.cpp [hunk 2]
```diff
@@ -31,9 +31,26 @@ void Scene::OnUpdate(const EngineContext &ctx) {
...
-    return PathFinding::FindPath(m_quadtree, start, dest, pathSize,
-                                 checkedOrders, detailLevel);
-}
+std::shared_ptr<NavMesh>
+Scene::BuildNavMesh(const Rect &bounds, const Vector2 &pathSize,
+                    float detailLevel, const PathValidator &validator) const {
+    const size_t width = static_cast<size_t>((bounds.Width() / detailLevel));
+    const size_t height = static_cast<size_t>((bounds.Height() / detailLevel));
+    const size_t gridSize = width * height;
+    float *grid = new float[gridSize];
```
[the remark this answers is not in the document]
Comment (joelsiks) [reply src/engine/Scene.cpp:40]: Sure, was just that it is named costs in the function it is passed to.

**Comment by joelsiks** [comment]
Good stuff

**Review by joelsiks — APPROVED** [review]
Good stuff!

**Pull request merged**